exporter/prometheusremotewriteexporter/wal.go

```diff
@@ -0,0 +1,246 @@
+	// https://docs.google.com/document/d/1cCcoFgjDFwU2n823tKuMvrIhzHty4UDyn0IcfUHiyyI/edit#heading=h.mlf37ibqjgov
+	// except that we are using underscores "_" instead of dashes "-".
+	Dir               string        `mapstructure:"directory"`
+	NBeforeTruncation int           `mapstructure:"cache_size"`
+	RefreshDuration   time.Duration `mapstructure:"truncate_frequency"`
+}
+
+func (wc *walConfig) createWAL() (*wal.Log, error) {
+	if wc == nil {
+		// There are cases for which the WAL can be disabled.
+		// TODO: Perhaps log that the WAL wasn't enabled.
+		return nil, nil
+	}
+	wal, err := wal.Open(filepath.Join(wc.Dir, "prom_rw"), &wal.Options{
+		SegmentCacheSize: wc.nBeforeTruncation(),
+		NoCopy:           true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("prometheusremotewriteexporter: failed to open WAL: %w", err)
+	}
+	return wal, nil
+}
+
+func (wc *walConfig) nBeforeTruncation() int {
+	if wc == nil {
+		return 0
+	}
+	if wc.NBeforeTruncation <= 0 {
+		return 300
+	}
+	return wc.NBeforeTruncation
+}
+
+func (wc *walConfig) refreshDuration() time.Duration {
+	if wc == nil {
+		return 0
+	}
+	if wc.RefreshDuration <= 0 {
+		// Following what Prometheus's WAL defaults.
+		return 1 * time.Minute
+	}
+	return wc.RefreshDuration
+}
+
+func (prwe *PrwExporter) walEnabled() bool { return prwe.walConfig != nil }
+
+var errWALDisabled = errors.New("WAL is disabled")
+
+func (prwe *PrwExporter) retrieveWALIndices(context.Context) (err error) {
+	if !prwe.walEnabled() {
+		return errWALDisabled
+	}
+
+	prwe.walMu.Lock()
+	defer prwe.walMu.Unlock()
+
+	wal, err := prwe.walConfig.createWAL()
+	if err != nil {
+		return err
+	}
+	prwe.wal = wal
+
+	prwe.rWALIndex, err = prwe.wal.FirstIndex()
+	if err != nil {
+		return fmt.Errorf("prometheusremotewriteexporter: failed to retrieve the last WAL index: %w", err)
+	}
+	prwe.wWALIndex, err = prwe.wal.LastIndex()
+	if err != nil {
+		return fmt.Errorf("prometheusremotewriteexporter: failed to retrieve the first WAL index: %w", err)
+	}
+	return nil
+}
+
+// exportThenTruncateWAL uploads the serialized data to the remote write endpoint,
+// and then truncates and retrieves the new indices for the WAL.
+func (prwe *PrwExporter) exportThenTruncateWAL(ctx context.Context, reqL []*prompb.WriteRequest) error {
+	if len(reqL) == 0 {
+		return nil
+	}
+	// Otherwise, time to flush.
+	if errs := prwe.exportWriteRequests(ctx, reqL); len(errs) != 0 {
+		return consumererror.Combine(errs)
+	}
+	// Save all the entries that aren't yet committed, to the tail of the WAL.
+	if err := prwe.wal.Sync(); err != nil {
+		return err
+	}
+	// Truncate the WAL from the front for the entries that we already
+	// read from the WAL and then exported.
+	if err := prwe.wal.TruncateFront(prwe.rWALIndex); err != nil && err != wal.ErrOutOfRange {
+		return err
+	}
+	// Reset by retrieving the respective read and write WAL indices.
+	return prwe.retrieveWALIndices(ctx)
```

Comment:
It's more like a question than a comment. My understanding is that `retrieveWALIndices` essentially reloads the WAL from disk and resets the read/write indexes accordingly. Why is that necessary? It seems somewhat heavy to me as it opens the file, etc. but I might miss something

Reply:
Basically there is one end that reads and then one that writes. This routine is invoked after we've flushed out to disk all prior entries. The idea is that afterwards, we want a true definition of where the read and write indexes are in the WAL and not just in RAM. It is needed otherwise the indices in RAM could be stale. This routine isn't invoked often and no it isn't heavy.

Reply:
Thanks @odeke-em for the explanation! There's just one thing I am unsure of - when the indices in RAM could go stale? Is there a valid scenario when this could happen or is it more like a precaution?

Reply:
When truncation of indices from the front occurs, it is up to the discretion of the tidwall code to change indices, hence we need to refresh.